developer_docs/codegen_stage.md

```diff
@@ -0,0 +1,281 @@
+```c
+t_ndarray f(t_ndarray a, t_ndarray b)
+{
+    int64_t i;
+    int64_t i_0001;
+    t_ndarray c = {.shape = NULL};
+    c = array_create(2, (int64_t[]){a.shape[INT64_C(0)], a.shape[INT64_C(1)]}, nd_int64, false);
+    for (i = INT64_C(0); i < c.shape[INT64_C(0)]; i += INT64_C(1))
+    {
+        for (i_0001 = INT64_C(0); i_0001 < c.shape[INT64_C(1)]; i_0001 += INT64_C(1))
+        {
+            GET_ELEMENT(c, nd_int64, (int64_t)i, (int64_t)i_0001) = GET_ELEMENT(a, nd_int64, (int64_t)i, (int64_t)i_0001) + GET_ELEMENT(b, nd_int64, (int64_t)i_0001);
+        }
+    }
+    return c;
+}
+```
+While in Fortran it becomes:
+```fortran
+  subroutine f(a, b, c)
+
+    implicit none
+
+    integer(i64), allocatable, intent(out) :: c(:,:)
+    integer(i64), intent(in) :: a(0:,0:)
+    integer(i64), intent(in) :: b(0:)
+    integer(i64) :: i
+
+    allocate(c(0:size(a, 1_i64, i64) - 1_i64, 0:size(a, 2_i64, i64) - &
+          1_i64))
+    do i = 0_i64, size(c, 2_i64, i64) - 1_i64, 1_i64
+      c(:, i) = a(:, i) + b
+    end do
+    return
+
+  end subroutine f
+```
+
+The loop unravelling function must be called every time we want to print a `CodeBlock` in a language which is less flexible than Python.
+The algorithm proceeds by inserting indices into the arrays at the dimensions furthest from the contiguous dimension.
+This is done recursively until the expression is compatible with the language.
+Where possible the function tries to group expressions in the same for loop.
+
+## C-specific problems
+
+### Use of C pointers
+
+In order to implement certain concepts in C it is important to use pointers.
+These concepts include returned variables and optional variables.
+This means that all print functions in the code must be able to correctly handle both pointers and non-pointers.
+In order to simply the developer's job and allow them to focus on the specifics of the AST element they are implementing, rather than the pointers `_print_Variable` and other similar print functions always print the value of the object.
+This is true even if the object is stored in a pointer (i.e. `(*var)` is printed).
+If an address is required then the object must be wrapped in the [`pyccel.ast.c_concepts.ObjectAddress`](../pyccel/ast/c_concepts.py) class.
+
+Although this construction ensures that valid code can be written easily, unless care is taken unidiomatic code will be produced (e.g. `(*var).shape` instead of `var->shape`).
+It is therefore very important to use the AST to represent class objects, so that the pointer differentiation can be handled in as few functions as possible.
+The following objects are very useful for this purpose:
+-   `DottedVariable` : Used to access member variables of a class. It may be necessary to create a `DottedVariable` to print an object such as a member of the `t_ndarray` class.
+-   `NumpyArraySize` : The size of an array in a given dimension
+-   `PyccelArraySize` : The total size of an array
```

Comment:
I agree. Our `NumpyArraySize` returns an element of the tuple returned by NumPy's `shape` function. Do you have a suggestion for a better name? If so I think we should create an issue. It is a small change but it will touch lots of parts of the code so I would prefer to fix this in a separate PR

Reply:
`NumpyArraySize`->`NumpyArrayShapeElement`
`PyccelArraySize`->`NumpyArraySize`
?

Reply:
Especially as every function where the name changes will want a new docstring

Reply:
I agree with your name suggestion. And yes, we need a separate issue for this.

Reply:
See #1329